contracts/Minter.sol

```diff
@@ -100,32 +104,62 @@ contract SPSMinter {
       token.mint(pools[i].receiver, amount);
 
       emit Mint(pools[i].receiver, amount);
+
+      unchecked { ++i; }
     }
   }
 
   /**
    * @notice Add new pool, can be called by admin
    * @param newReceiver Address of the receiver
    * @param newAmount Amount of tokens per block
+   * @param newReductionBlocks Number of blocks between emission reduction
+   * @param newReductionPercentage Percentage to reduce emission
    */
-  function addPool(address newReceiver, uint256 newAmount) external onlyAdmin {
+  function addPool(address newReceiver, uint256 newAmount, uint256 newReductionBlocks, uint256 newReductionPercentage) external onlyAdmin {
     require(pools.length < poolsCap, 'SPSMinter: Pools cap reached');
     require(newAmount <= maxToPoolPerBlock, 'SPSMinter: Maximum amount per block reached');
-    pools.push(Pool(newReceiver, newAmount));
-    emit PoolAdded(newReceiver, newAmount);
+    pools.push(Pool(newReceiver, newAmount, newReductionBlocks, newReductionPercentage, block.number));
+    emit PoolAdded(newReceiver, newAmount, newReductionBlocks, newReductionPercentage, block.number);
   }
 
   /**
    * @notice Update pool, can be called by admin
    * @param index Index in the array of the pool
    * @param newReceiver Address of the receiver
    * @param newAmount Amount of tokens per block
+   * @param newReductionBlocks Number of blocks between emission reduction
+   * @param newReductionPercentage Percentage to reduce emission
    */
-  function updatePool(uint256 index, address newReceiver, uint256 newAmount) external onlyAdmin {
+  function updatePool(uint256 index, address newReceiver, uint256 newAmount, uint256 newReductionBlocks, uint256 newReductionPercentage) external onlyAdmin {
     require(newAmount <= maxToPoolPerBlock, 'SPSMinter: Maximum amount per block reached');
     mint();
-    pools[index] = Pool(newReceiver, newAmount);
-    emit PoolUpdated(index, newReceiver, newAmount);
+    pools[index] = Pool(newReceiver, newAmount, newReductionBlocks, newReductionPercentage, block.number);
+    emit PoolUpdated(index, newReceiver, newAmount, newReductionBlocks, newReductionPercentage, block.number);
+  }
+
+  /**
+   * @notice Update emissions for pools
+   * @param index Index in the array of the pool
+   * @param updateAll If true, all pools will be updated
+   */
+  function updateEmissions(uint256 index, bool updateAll) public {
+    if (updateAll){
+      uint256 length = pools.length;
+      for (uint256 i = 0; i < length;){
+        if (block.number - pools[i].lastUpdate > pools[i].reductionBlocks){
+          pools[i].amountPerBlock = pools[i].amountPerBlock / 100 * (100 - pools[i].reductionPercentage);
```

Comment:
You want to make sure you do multiplication before division to maximize precision:

```solidity
(pools[i].amountPerBlock * (100 - pools[i].reductionPercentage)) / 100;
```

----

Also, as written, this will keep decaying forever until it reaches 0 due to integer division. This will mean there is a long period of time during which you'll wasting gas in `mint` giving pools tiny token amounts, worth only a few cents. It would probably be better to pick a threshold (1 token? 0.5 tokens? 0.1 tokens?) at which you'll intentionally round down to 0. This can be a contract constant somewhere.

Reply:
Resolved, I used 0.1 tokens as a minimum, after that, it's rounded to 0: https://github.com/steem-monsters/sps-minter/commit/7eb68258dda51e0e813436bafa2651c6b399c712, https://github.com/steem-monsters/sps-minter/commit/059a28c60e4c750af8e8d1bf6599abf89f032be5